### PR TITLE
BI-2255 - Experimental Collaborator Micronaut Changes

### PR DESCRIPTION
**Pull request by davedrp**
# Description
[BI-2255](https://breedinginsight.atlassian.net/browse/BI-2255) - Experimental Collaborator Micronaut Changes

- Added `EXPERIMENTAL_COLLABORATOR` to the enum `ProgramSecuredRole`
- Changed `ProgramSecuredRoleGroup.ALL` to `ProgramSecuredRoleGroup.PROGRAM_SCOPED_ROLES`.
- Added `ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR` to `ProgramSecuredRoleGroup.PROGRAM_SCOPED_ROLES`
- Created the `@ExperimentCollaboratorSecured` annotation for endpoints with experimentId (or traitId)

# Testing
see Acceptance Criteria in [BI-2255](https://breedinginsight.atlassian.net/browse/BI-2255)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [ ] I have tested that my code works with both the brapi-java-server and BreedBase
- [ ] I have create/modified unit tests to cover this change
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_

[BI-2255]: https://breedinginsight.atlassian.net/browse/BI-2255?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/org/breedinginsight/api/auth/rules/ExperimentCollaboratorSecuredAnnotationRule.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        String experimentId = (String) routeMatch.getVariableValues()
+                .get("experimentId");
+        if( experimentId==null) {
+            experimentId = (String) routeMatch.getVariableValues()
+                    .get("trialId");
+        }
+        return experimentId;
+    }
+
+    private SecurityRuleResult processExperiment(AuthenticatedUser authenticatedUser, String experimentId, String programId) {
```
Comment (mlm483) [inline src/main/java/org/breedinginsight/api/auth/rules/ExperimentCollaboratorSecuredAnnotationRule.java:122]: Could you rename this method? We use the term "process" in a lot of our import code, and I think for that reason it is confusing to use it here. Something like `checkAuthorization` or `getSecurityRuleResult` would be more clear.

File: src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java
Comment (mlm483) [inline src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java:1]: The `getProgram` and `getPrograms` methods in this controller should be annotated with 
```java
    @ProgramSecured(roles = {ProgramSecuredRole.SYSTEM_ADMIN, ProgramSecuredRole.PROGRAM_ADMIN, ProgramSecuredRole.READ_ONLY, ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR})
```

File: src/main/java/org/breedinginsight/brapi/v2/BrAPIStudiesController.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ public BrAPIStudiesController(BrAPIStudyService studyService, StudyQueryMapper s
 
     @Get("/studies{?queryParams*}")
     @Produces(MediaType.APPLICATION_JSON)
-    @ProgramSecured(roleGroups = {ProgramSecuredRoleGroup.ALL})
+    @ProgramSecured(roleGroups = {ProgramSecuredRoleGroup.PROGRAM_SCOPED_ROLES})
     public HttpResponse<Response<DataResponse<List<BrAPIStudy>>>> getStudies(
```
Comment (mlm483) [inline src/main/java/org/breedinginsight/brapi/v2/BrAPIStudiesController.java:78]: Ultimately this method needs to allow Experimental Collaborator role access. You can either do that in this PR or I will make the change in BI-2258 (where I have also implemented filtering in this method).

File: src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java [hunk 3]
```diff
@@ -38,19 +39,21 @@
...
         this.experimentService = experimentService;
         this.experimentQueryMapper = experimentQueryMapper;
         this.referenceSource = referenceSource;
     }
 
     @Get("/trials{?queryParams*}")
     @Produces(MediaType.APPLICATION_JSON)
-    @ProgramSecured(roleGroups = {ProgramSecuredRoleGroup.ALL})
+    @ProgramSecured(roleGroups = {ProgramSecuredRoleGroup.PROGRAM_SCOPED_ROLES})
     public HttpResponse<Response<DataResponse<List<BrAPITrial>>>> getExperiments(
```
Comment (mlm483) [inline src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java:57]: Needs to allow Experimental Collaborator role access. You could do that in the PR or else I'll do it in BI-2258 where I've implemented filtering in this method.

**Review by mlm483 — CHANGES_REQUESTED** [review]
I haven't tested functionality yet, I wanted to get my change requests in ASAP.

See comments for changes, the most important of which is:
- `ProgramSecuredRoleGroup.PROGRAM_SCOPED_ROLES` currently includes `ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR` and it should not.

File: src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java [hunk 4]
```diff
@@ -191,6 +191,7 @@ public HttpResponse<Response<ImportMapping>> editMapping(@PathVariable UUID prog
     @Produces(MediaType.APPLICATION_JSON)
     @AddMetadata
     @Secured(SecurityRule.IS_ANONYMOUS)
+    @ProgramSecured(roles = {ProgramSecuredRole.PROGRAM_ADMIN, ProgramSecuredRole.SYSTEM_ADMIN})
     public HttpResponse<Response<DataResponse<ImportMapping>>> getSystemMappings(@Nullable @QueryValue String importName) {
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java:195]: It also looks like the unit test is failing

File: src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java [hunk 5]
```diff
@@ -117,4 +121,11 @@ private void setDbIds(BrAPITrial trial) {
         //TODO update locationDbId
     }
 
+    private boolean isExperimentCoordinator(ProgramUser programUser){
+        List<Role> roles = programUser.getRoles();
+        return (roles.size()==1 &&
+                ProgramSecuredRole.getEnum(roles.get(0).getDomain())==ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR);
+
+    }
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java:129]: Looks like it was probably moved to ExperimentCollaboratorSecuredAnnotationRule and updated

**Review by nickpalladino — CHANGES_REQUESTED** [review]
I'm seeing some integration tests failing that will need to be resolved. Some of them might be fixed by some of the suggested changes in this PR.

File: src/main/java/org/breedinginsight/api/auth/ExperimentCollaboratorSecuredRole.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.breedinginsight.api.auth;
+
+public enum ExperimentCollaboratorSecuredRole {
+    EXPERIMENTAL_COLLABORATOR("Experimental Collaborator");
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/api/auth/ExperimentCollaboratorSecuredRole.java]: done.

File: src/main/java/org/breedinginsight/api/auth/ProgramSecuredRoleGroup.java [hunk 7]
```diff
@@ -17,13 +17,10 @@
...
 package org.breedinginsight.api.auth;
 
-import org.apache.commons.collections4.ListUtils;
-
 import java.util.List;
 
 public enum ProgramSecuredRoleGroup {
-    ALL_PROGRAM_ROLES(List.of(ProgramSecuredRole.READ_ONLY, ProgramSecuredRole.PROGRAM_ADMIN)),
-    ALL(ListUtils.union(ALL_PROGRAM_ROLES.getProgramRoles(), List.of(ProgramSecuredRole.SYSTEM_ADMIN)));
+    PROGRAM_SCOPED_ROLES(List.of(ProgramSecuredRole.SYSTEM_ADMIN, ProgramSecuredRole.READ_ONLY, ProgramSecuredRole.PROGRAM_ADMIN, ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR));
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/api/auth/ProgramSecuredRoleGroup.java]: Thanks, I now understand the new role.
The recommended changes have been made.

File: src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java [hunk 8]
```diff
@@ -84,7 +84,7 @@ public HttpResponse<Response<DataResponse<ImportConfigResponse>>> getImportTypes
     @Get("/programs/{programId}/import/mappings{?draft}")
     @Produces(MediaType.APPLICATION_JSON)
     @AddMetadata
-    @ProgramSecured(roles = {ProgramSecuredRole.PROGRAM_ADMIN, ProgramSecuredRole.SYSTEM_ADMIN})
+    @ProgramSecured(roles = {ProgramSecuredRole.PROGRAM_ADMIN, ProgramSecuredRole.SYSTEM_ADMIN, ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR})
     public HttpResponse<Response<DataResponse<ImportMapping>>> getMappings(@PathVariable UUID programId,
                                                                            @QueryValue(defaultValue = "false") Boolean draft) {
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java]: The code has been reverted.

File: src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapps/importer/controllers/ImportController.java]: @nickpalladino yes it was causing a failure

File: src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java [hunk 9]
```diff
@@ -38,19 +39,21 @@
...
     private final ExperimentQueryMapper experimentQueryMapper;
 
     @Inject
-    public BrAPITrialsController(BrAPITrialService experimentService, ExperimentQueryMapper experimentQueryMapper, @Property(name = "brapi.server.reference-source") String referenceSource) {
+    public BrAPITrialsController(SecurityService securityService, BrAPITrialService experimentService, ExperimentQueryMapper experimentQueryMapper, @Property(name = "brapi.server.reference-source") String referenceSource) {
+        this.securityService = securityService;
         this.experimentService = experimentService;
         this.experimentQueryMapper = experimentQueryMapper;
         this.referenceSource = referenceSource;
     }
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java]: `SecurityService` is not used.  It has been removed.

File: src/main/java/org/breedinginsight/api/auth/rules/ExperimentCollaboratorSecuredAnnotationRule.java (hunk 1 above)
Comment (davedrp) [reply src/main/java/org/breedinginsight/api/auth/rules/ExperimentCollaboratorSecuredAnnotationRule.java]: renamed ` checkAuthorization`

File: src/main/java/org/breedinginsight/brapi/v2/BrAPIStudiesController.java (hunk 2 above)
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapi/v2/BrAPIStudiesController.java]: The `roles` now includes `ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR`

File: src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java
Comment (davedrp) [reply src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java:1]: The `roles` now includes `ProgramSecuredRole.EXPERIMENTAL_COLLABORATOR`

File: src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java]: removed

File: src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java [hunk 10]
```diff
@@ -88,6 +88,8 @@ public ProgramController(ProgramService programService, ProgramUserService progr
 
     @Get("/programs{?queryParams*}")
     @Produces(MediaType.APPLICATION_JSON)
+    @ProgramSecured(roles = {ProgramSecuredRole.SYSTEM_ADMIN, ProgramSecuredRole.READ_ONLY, ProgramSecuredRole.PROGRAM_ADMIN
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java:91]: `ProgramSecuredAnnotationRuleIntegrationTest::getAllProgramsAdmin`,`ProgramSecuredAnnotationRuleIntegrationTest::getAllProgramsMember`, `ProgramControllerIntegrationTest::getProgramsSuccess`, `ProgramControllerIntegrationTest::getProgramQuery` failing with error message `Endpoint does not have program id to check roles against` with annotation added on this method.

**Review by nickpalladino — CHANGES_REQUESTED** [review]

**Review by dmeidlin — CHANGES_REQUESTED** [review]
When navigating to the system Programs section the back end returns a 500 response to GET calls to /v1/programs, resulting in no programs being listed.
ERROR o.b.a.v.c.InternalServerErrorHandler - Error Id: 91a2d7e4-f1a9-4527-906e-5caa9a9d0eb3
io.micronaut.http.server.exceptions.HttpServerException: Endpoint does not have program id to check roles against
	at org.breedinginsight.api.auth.rules.ProgramSecuredAnnotationRule.check(ProgramSecuredAnnotationRule.java:66)

File: src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java (hunk 10 above)
Comment (davedrp) [reply src/main/java/org/breedinginsight/api/v1/controller/ProgramController.java:91]: Fixed

**Comment by davedrp** [comment]
> When navigating to the system Programs section as a SysAdmin the back end returns a 500 response to GET calls to /v1/programs, resulting in no programs being listed. ERROR o.b.a.v.c.InternalServerErrorHandler - Error Id: 91a2d7e4-f1a9-4527-906e-5caa9a9d0eb3 io.micronaut.http.server.exceptions.HttpServerException: Endpoint does not have program id to check roles against at org.breedinginsight.api.auth.rules.ProgramSecuredAnnotationRule.check(ProgramSecuredAnnotationRule.java:66) Same error also when loggin in as Program Admin and Read Only.

Fixed

**Review by nickpalladino — APPROVED** [review]

**Review by mlm483 — APPROVED** [review]

**Review by dmeidlin — APPROVED** [review]

**Pull request merged**